services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbCallExecutor;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDocumentBuilder;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.Versioned;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.Result;
+import org.eclipse.hono.service.management.tenant.Tenant;
+import org.eclipse.hono.service.management.tenant.TenantManagementService;
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.AuthenticationConstants;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import com.mongodb.MongoCommandException;
+
+import io.opentracing.Span;
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.UpdateOptions;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+public final class MongoDbBasedTenantService extends AbstractVerticle
+        implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private static final int INDEX_CREATION_MAX_RETRIES = 3;
+    private final String FIELD_PAYLOAD_TENANT = "tenant";
+    private final String FIELD_PAYLOAD_TRUSTED_CA_WITH_SUBJECT_DN = FIELD_PAYLOAD_TENANT + "."
+            + RegistryManagementConstants.FIELD_PAYLOAD_TRUSTED_CA + "."
+            + AuthenticationConstants.FIELD_SUBJECT_DN;
+    private final String DUPLICATE_ERROR_MONGODB_REGEX = "^\\w+\\sduplicate key error collection:\\s.+\\sindex:\\s.+\\sdup\\skey:\\s\\{\\s(?<key>.+):\\s(?<val>.+)\\s}$";
+    private final Pattern duplicateErrorMongodbPattern = Pattern.compile(DUPLICATE_ERROR_MONGODB_REGEX);
```

Comment:
IMHO I wouldn't rely on parsing the error message and make a decision based on that. As we discussed, in case of duplicate tenant ID or _subjectDN_ `HTTP_CONFLICT` is returned. In that case I would simply customise the log message to _The tenant with the given identifier already exists or an existing tenant uses a certificate authority with the same Subject DN_ as given in the [API docs](https://www.eclipse.org/hono/docs/api/management/#/tenants/createTenantWithId). WDYT?

Reply:
done